[online_image] Add binary bmp support

jesserockz opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

This adds basic support for binary BMP images to be downloaded and displayed. This PR will **not** add more functionality. If anyone wants colour support, it can be done in a further PR after this is merged.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4600

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

jesserockz committed:
[online_image] Add binary bmp support

jesserockz committed:
Add to tests

jesserockz committed:
Merge branch 'dev' into jesserockz-2025-014

jesserockz committed:
Fix variable names

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+    // Check if the file is a BMP image
+    if (buffer[0] != 'B' || buffer[1] != 'M') {
+      ESP_LOGE(TAG, "Not a BMP file");
+      return 0;
```
Errors in online_image decoders are normally signaled with negative values
```suggestion
      return -1;
```

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+
+int HOT BmpDecoder::decode(uint8_t *buffer, size_t size) {
+  size_t index = 0;
+  this->decoded_bytes_ += size;
```
It somehow feels wrong to increase decoded_bytes before actually decoding them 🤔

jesserockz replied:
I don't disagree. I can rework this

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+      return 0;
+    }
+
+    this->total_size_ = encode_uint32(buffer[5], buffer[4], buffer[3], buffer[2]);
```
Might be nice to add some comments about what the magic numbers represent / how the header in the file is formatted

jesserockz replied:
I can add something, but isn't the variable name the explanation/description of what is in the (4) bytes?

guillempages replied:
It's just a reflex. I can't stand "magic numbers", but I agree that defining constants here everywhere for the indexes would be overkill.
But having a nice comment with an explanation about the format and how it works might help later on, when implementing additional compression methods or bpps.

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+    this->current_index_ = 14;
+    index = 14;
+  }
+  if (this->current_index_ == 14 && this->decoded_bytes_ > data_offset_) {
```
```suggestion
  if (this->current_index_ == 14 && this->decoded_bytes_ > this->data_offset_) {
```

guillempages commented on esphome/components/online_image/online_image.cpp:
```diff
@@ -120,9 +124,14 @@ void OnlineImage::update() {
-    this->decoder_ = esphome::make_unique<PngDecoder>(this);
-  }
+    this->decoder_ = make_unique<PngDecoder>(this);
+  } else
```
```suggestion
  }
```
this->format can only have one value, so it should behave exactly the same without an "else" chain.

With the #ifdef combination it might cause issues with the else not having a block, depending on which combination of defines is actually compiled in.

This is also only called once per image download, so the extra few cycles that the additional if's would cost should be negligible.

guillempages commented on esphome/components/online_image/online_image.cpp:
```diff
@@ -120,9 +124,14 @@ void OnlineImage::update() {
+  } else
 #endif  // ONLINE_IMAGE_PNG_SUPPORT
+#ifdef USE_ONLINE_IMAGE_BMP_SUPPORT
+      if (this->format_ == ImageFormat::BMP) {
```
```suggestion
  if (this->format_ == ImageFormat::BMP) {
```
If the else is removed, no need to indent this further

jesserockz replied:
That was the formatter =)

guillempages commented on tests/components/online_image/common.yaml:
```diff
@@ -26,6 +26,10 @@ online_image:
     type: RGB
     transparency: chroma_key
+  - id: online_binary_bmp
+    url: http://www.libpng.org/pub/png/img_png/pnglogo-blk-tiny.png
```
It's not really executed, so it's not that important, but it would feel more consistent, if the URL would point to a bmp image ;-)

jesserockz replied:
Haha yes. I just didnt know any and was being lazy at the end.

guillempages replied:
```suggestion
    url: https://samples-files.com/samples/images/bmp/480-360-sample.bmp
```
As mentioned, feel free to ignore (nobody knows, how long the link will be valid); but it feels better to have an actual BMP link ;-)

guillempages commented on esphome/components/online_image/bmp_image.h:
```diff
@@ -0,0 +1,41 @@
+
+ protected:
+  size_t current_index_{0};
+  ssize_t width_{0};
```
what's the difference between `size_t` (above) and `ssize_t` (here)?

jesserockz replied:
ssize_t is signed size_t, the BMP format allows for negative numbers, mostly negative height which indicates a top down image vs the standard bottom up image.

guillempages commented on esphome/components/online_image/bmp_image.h:
```diff
@@ -0,0 +1,41 @@
+ protected:
+  size_t current_index_{0};
+  ssize_t width_{0};
+  ssize_t height_{0};
```
Are width and height really necessary here? they are already used (and set) in the destination image; can't those be used instead?

jesserockz replied:
They are required to know the x and y of the image we are drawing, for example if someone tried to download an image larger than the "container" it will attempt to draw outside the bounds (which is just ignored and logged)

guillempages replied:
Actually this out-of-bounds checking should already be checked in the draw method in the ImageDecoder superclass. It even takes care of resizing, if the container is smaller than the image.
So using "setSize()" and the actual image width and height should be ok.

guillempages commented on esphome/components/online_image/bmp_image.h:
```diff
@@ -0,0 +1,41 @@
+  size_t current_index_{0};
+  ssize_t width_{0};
+  ssize_t height_{0};
+  uint16_t bits_per_pixel_{0};
```
Same here; the destination image already should have/get this information. Or is this about the "source" bpp? Is there any conversion happening if the destination image uses a different format?

jesserockz replied:
This is decoded in the first iteration from the header, but will be used later when more bpp are supported to know how to decode the individual bytes which happens in later loops where a local variable wont exist.

jesserockz replied:
No conversion in this code, but it would be possible in a future PR to say I want greyscale, but the source is RGB etc.

guillempages commented on esphome/components/online_image/bmp_image.h:
```diff
@@ -0,0 +1,41 @@
+  uint32_t image_data_size_{0};
+  uint32_t color_table_entries_{0};
+  size_t width_bytes_{0};
+  size_t total_size_{0};
```
what's the difference to the "download_size_" of the superclass? Can't that one be used/updated?

jesserockz replied:
I don't believe they would be any different, We could probably update that. If a server doesnt present a Content-Length, then this variable from the header would server as a good replacement of the size of the entire file.

jesserockz replied:
I have removed `total_size_` and assigned the value read from the header to `download_size_`

guillempages commented on esphome/components/online_image/bmp_image.h:
```diff
@@ -0,0 +1,41 @@
+class BmpDecoder : public ImageDecoder {
+ public:
+  /**
+   * @brief Construct a new PNG Decoder object.
```
```suggestion
   * @brief Construct a new BMP Decoder object.
```

jesserockz replied:
:see_no_evil:

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+        break;
+      default:
+        ESP_LOGE(TAG, "Unsupported bits per pixel: %d", this->bits_per_pixel_);
+        return 0;
```
```suggestion
        return -2;
```
Maybe we should define some constants for possible errors in the "ImageDecoder" class

jesserockz replied:
I like this idea

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+static const char *const TAG = "online_image.bmp";
+
+int HOT BmpDecoder::decode(uint8_t *buffer, size_t size) {
+  size_t index = 0;
```
I might have missed that, but it looks like `index` is not really used (besides being set at different places), and only `this->current_index_` is actually being read.

Is this variable really needed?

jesserockz replied:
It is used to know which byte in the current buffer it is up to.
https://github.com/esphome/esphome/pull/8116/files#diff-6400dadccb239aaadf18aad140f9f795cb6dc755e1b05171203a161b09da779fR56

jesserockz replied:
It is now used to loop the buffer instead of current_index (which is used to get the paint_index)

guillempages reviewed (commented):
Adding only partial support for (binary) BMP might cause false expectations if people then start wildly downloading BMPs.
OTOH, it's a start, and additional BMP encodings/formats can be added later.

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    if (!this->set_size(this->width_, this->height_)) {
+      return 0;
```
```suggestion
      return -3;
```

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -0,0 +1,72 @@
+        ESP_LOGE(TAG, "Unsupported bits per pixel: %d", this->bits_per_pixel_);
+        return 0;
+    }
+
```
Maybe also return an error "unsupported format" if the compression_method is not "uncompressed"?

jesserockz committed:
Apply suggestions from code review

Co-authored-by: guillempages <guillempages@users.noreply.github.com>

jesserockz committed:
Add comments
Use index instead of current_index for buffer iteration
Return error on unsupported compression method

jesserockz committed:
Fix enum name

guillempages commented on esphome/components/online_image/bmp_image.cpp:
```diff
@@ -13,21 +13,42 @@ static const char *const TAG = "online_image.bmp";
+     * 2-5: File size
+     * 6-9: Reserved
+     * 10-13: Pixel data offset
+     */
```
```suggestion
     *
     * Integer values are stored in little-endian format.
     */
```
Probably obvious to anyone working with BMP, but IMHO it doesn't hurt to point this out ;-)

guillempages commented on esphome/components/online_image/image_decoder.h:
```diff
@@ -4,6 +4,12 @@
 namespace esphome {
 namespace online_image {
 
+enum DecodeError : int {
```
Hm. I think the project maintainers might consider this to be worthy of a separate PR. Jesse often mentions that smaller separate PRs are easier to review 😜 

Jokes apart, this looks good. Once this is merged I'll update the PNG decoder to make use of those values as well.

jesserockz replied:
:rofl:

guillempages reviewed (commented):
Looking good.
Just a final set of suggestions.

jesserockz committed:
Apply suggestions from code review

Co-authored-by: guillempages <guillempages@users.noreply.github.com>

guillempages reviewed (approved):
LGTM
